src/tex/crunch_texture.cpp

```diff
@@ -0,0 +1,192 @@
+/* Copyright (C) 2021 Edgar B
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include <btu/common/string.hpp>
+#include <btu/tex/crunch_texture.hpp>
+#include <btu/tex/dimension.hpp>
+#include <btu/tex/formats.hpp>
+#include <crunch/crn_cfile_stream.h>
+#include <crunch/crn_core.h>
+#include <crunch/crn_dynamic_stream.h>
+#include <crunch/crn_image.h>
+#include <crunch/crn_mipmapped_texture.h>
+#include <winerror.h>
+
+#include <mutex>
+
+namespace crnlib {
+auto operator==(const image_u8 &lhs, const image_u8 &rhs) noexcept -> bool
+{
+    // NOLINTBEGIN(cppcoreguidelines-pro-bounds-pointer-arithmetic)
+    auto *lhs_end = lhs.get_ptr() + lhs.get_total();
+    auto *rhs_end = rhs.get_ptr() + rhs.get_total();
+    // NOLINTEND(cppcoreguidelines-pro-bounds-pointer-arithmetic)
+    return std::equal(lhs.get_ptr(), lhs_end, rhs.get_ptr(), rhs_end);
+}
+
+auto operator==(const mipmapped_texture &lhs, const mipmapped_texture &rhs) noexcept -> bool
+{
+    auto cmp = [&](auto... ptrs) { return ((std::invoke(ptrs, lhs) == std::invoke(ptrs, rhs)) && ...); };
+    const bool first = cmp(&mipmapped_texture::get_num_faces,
+                           &mipmapped_texture::get_num_levels,
+                           &mipmapped_texture::get_total_pixels);
+    if (!first)
+        return false;
+
+    image_u8 lhs_temp, rhs_temp;
+    for (uint face = 0; face < lhs.get_num_faces(); face++)
+    {
+        for (uint level = 0; level < lhs.get_num_levels(); level++)
+        {
+            image_u8 *lhs_img = lhs.get_level_image(face, level, lhs_temp);
+            image_u8 *rhs_img = rhs.get_level_image(face, level, rhs_temp);
+            if (*lhs_img != *rhs_img)
+                return false;
+        }
+    }
+
+    return true;
+}
+} // namespace crnlib
+
+namespace btu::tex {
+using namespace crnlib;
+
+void CrunchTexture::set(mipmapped_texture &&tex) noexcept
+{
+    tex_ = std::move(tex);
+}
+
+auto CrunchTexture::get() noexcept -> mipmapped_texture &
+{
+    return tex_;
+}
+
+auto CrunchTexture::get() const noexcept -> const mipmapped_texture &
+{
+    return tex_;
+}
+
+auto CrunchTexture::get_dimension() const noexcept -> Dimension
+{
+    return {tex_.get_width(), tex_.get_height()};
+}
+
+auto CrunchTexture::get_load_path() const noexcept -> const Path &
+{
+    return load_path_;
+}
+
+void CrunchTexture::set_load_path(Path path) noexcept
+{
+    load_path_ = std::move(path);
+}
+
+auto CrunchTexture::get_texture_type() const noexcept -> const TextureType
+{
+    const auto filename = load_path_.filename().u8string();
+
+    const auto guessed = guess_texture_type(filename);
+    if (!guessed)
+        return TextureType::Diffuse;
+    else
+        return guessed.value();
+}
```

Comment:
You can return the optional directly. The caller can then do `guessed.value_or(TextureType::Diffuse)`

Reply:
Hmm, makes this function a bit redundant. It's a bit more user friendly though. What about using `return guessed.value_or(TextureType::Diffuse);` instead of the `if` then?

Reply:
I think letting the caller do it is more explicit. The function name does not imply a default value could be used
There might be some cases where the caller prefers a different default value

Alternatively, you can rename the function and maybe pass the default value in parameter, defaulting to diffuse